### DISCUSSION TITLE
How to prevent execution of entry actions after reading persisted state?

### DISCUSSION THREAD

**Discussion by bennyKY** [statelyai/xstate#1732] [General] [answered]
Hi, after persisting a state:
```javascript
JSON.stringify(state)
```

and then reading it:
```javascript
const previousState = State.create(JSON.parse(userState._state));
resolvedState = machine.resolveState(previousState);
interpret(machine).start(resolvedState);
```

It runs the `entry` actions on this state, but i just saved it and obviously these actions already completed.
So i dont need to run these actions again after loading the state.
Is there any proper solution?

**Comment by davidkpiano** [comment, marked as answer]
You can remove the actions from the resolved state:
```js
resolvedState.actions.length = 0; // remove actions
interpret(machine).start(resolvedState);
```

**Reply by bennyKY** [reply]
Thanks a lot!
Does it has any implications on the `exitActions` or any other actions?

**Reply by davidkpiano** [reply]
This will remove all the actions only on the state passed in. Any future actions (from other state transitions) will still be executed.

**Reply by bennyKY** [reply]
Thanks a lot!
You did a great job with this library. We are using it in production in Node environment (which i think much more rare than using it in frontend).

**Reply by davidkpiano** [reply]
> We are using it in production in Node environment (which i think much more rare than using it in frontend).

That's awesome! You'd be surprised how much XState is used in the backend, as well. I'm working on tools to make that experience even better.

**Reply by jirutka** [reply]
@davidkpiano, can you please share the tools, examples or any other resources about using XState in backend? I’m currently trying to integrate XState as a workflow engine for business processes. I’m starting to be really desperate because **all** the examples I can find are for front-end use cases and/or too trivial, so not very useful for me. There are too many options to do something and I just cannot figure out which ones are the right or optimal.

**Reply by davidkpiano** [reply]
@jirutka Sure: https://github.com/davidkpiano/durable-entities-xstate

**Reply by bennyKY** [reply]
@davidkpiano Hi, now we want to add `exitActions`, but as you said here:
> This will remove all the actions only on the state passed in. Any future actions (from other state transitions) will still be executed.

It wont execute them because we are removing all actions by:
```javascript
resolvedState.actions.length = 0;
```
So the requirements are :
1. Not execute entry actions when starting the machine after persisting it (because it was already executed upon the event, before persisting it),
2. Execute exit actions after reading the persisted state upon state transition.


Doing `resolvedState.actions.length = 0;` its a hack, is it really the only way? if it is, what do you think is the best option to keep the `exitActions` in place?

**Reply by davidkpiano** [reply]
Actions are not discriminated; actions are actions, whether they're entry, exit, transition, etc. What is your use-case for wanting to keep exit actions in place?
